sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1272,7 +1272,8 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    * Create a [[Cast]] expression.
    */
   override def visitCast(ctx: CastContext): Expression = withOrigin(ctx) {
-    Cast(expression(ctx.expression), visitSparkDataType(ctx.dataType))
+    Cast(expression(ctx.expression()), visitSparkDataType(ctx.dataType),
```

Comment:
`Cast` is pretty popular expression. Even it is not a part of public API, by changing its signature we can break some apps. How about **new** expression like `Cast` which accepts `format` as a field. The expression case class could inherit most of `Cast`.

Reply:
As far as I know case class (the new expression `FormattedCast`) can't inherit from another case class (`Cast`). We could create a `CastBase` trait or abstract class with the common logic and inherit inherit both `Cast` and `FormattedCast` from it, but I still see the issue that many places we use pattern matching with `Cast`. We can add matching to `FormattedCast` to these places now, but I feel it will be a bit error prone for the future and some might forget to match `FormattedCast` too.

Reply:
Last time when I changed `Cast`, @cloud-fan had some concerns about it, see https://github.com/apache/spark/pull/24342#discussion_r276112428 . Personally I am ok to add new field to `Cast` but let's ask other opinions. /cc @srowen @gatorsmile @HyukjinKwon

Reply:
Technically it's fine to change - otherwise, we will be stuck every time we fix something even internal.

In practice, yes, some people use those internal catalyst APIs (as Max pointed out, it's likely Cast is being used somewhere else). So, it's better to avoid change it possible.

So, let's change when it's absolutely needed. I haven;t taken a close look yet but let's change it if this is something inevitable.